### ISSUE TITLE
Is it possible to get exact value of attention probability or logit?

### ISSUE THREAD

**Issue by ojm9898** [HazyResearch/flash-attention#106]
Hello Tri,
Thank you for your great work!

When I adjust FlashAttention with Copy mechanism(e.g. [pointer-generator](https://arxiv.org/abs/1704.04368))
I need attention probability or attention score(before scaling) to calculate copy loss.

I did like this,

        context_layer, softmax_lse, softmax_output = flash_attn_unpadded_func(q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
                                                                              self.dropout_rate if self.training else 0.0,
                                                                              softmax_scale=None,
                                                                              causal=causal,
                                                                              return_attn_probs=True)  # [total_t, n_heads, head_size], [batch_size, n_heads, seqlen_q], [batch_size, n_heads, seqlen_q, seqlen_k]
        attention_score = torch.exp(softmax_lse[:, :, :max_seqlen_q, None]) * softmax_output[:, :, :max_seqlen_q, :max_seqlen_k]  # [batch_size, n_heads, seqlen_q, seqlen_k]
        attention_score = attention_score.mean(dim=1)  # [batch_size, seqlen_q, seqlen_k], head dimension will be mixed.

but attention score is not accurate and there are so many zero elements.

Is it possible to get attention probability or logit by inverce calculating the output of flash_attn_func?

Thanks.
Jumin

**Comment by tridao** [comment]
By default we don't store the attention score/probability, since that's a large matrix that makes attention slow and memory-hungry. 
If you need the attention score/probability, it's better to just use a standard implementation of attention.

The `return_attn_probs` option is for internal testing only, as mentioned in our docstring.
```
        return_attn_probs: bool. Whether to return the attention probabilities. This option is for
           testing only. The returned probabilities are not guaranteed to be correct
           (they might not have the right scaling).
```
If you really want to see how we use this returned probabilities to get the right attention probability, you can take a look at our [test](https://github.com/HazyResearch/flash-attention/blob/780e8eeabb84fe3f41e8244f04521743b032ba35/tests/test_flash_attn.py#L387).
But again, this is for testing only and it's easier/faster to get the attention probability by using a standard implementation.

**Comment by ojm9898** [comment]
Thank you for your quick reply. Have a nice day!
